From 2842eda1a760876c0b2c185b99ef1613955e4124 Mon Sep 17 00:00:00 2001
From: Myron Stowe <myron.stowe@redhat.com>
Date: Thu, 28 Aug 2014 03:06:20 -0400
Subject: [acpi] dock: Use acpi_device_enumerated() to check if dock is present

Message-id: <20140828030620.2117.77815.stgit@gir.stowe>
Patchwork-id: 89315
O-Subject: [RHEL7.1 PATCH 27/30] ACPI / dock: Use acpi_device_enumerated() to check if dock is present
Bugzilla: 1128632
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>
RH-Acked-by: Stefan Assmann <sassmann@redhat.com>

Bugzilla: https://bugzilla.redhat.com/show_bug.cgi?id=1128632
Upstream Status: 0a8e5c3d5f0f4929761e6a5bef5358f0ccd8810c

commit 0a8e5c3d5f0f4929761e6a5bef5358f0ccd8810c
Author: Rafael J. Wysocki <rafael.j.wysocki@intel.com>
Date:   Mon Feb 10 13:44:20 2014 +0100

    ACPI / dock: Use acpi_device_enumerated() to check if dock is present

    After commit 202317a573b2 (ACPI / scan: Add acpi_device objects for
    all device nodes in the namespace) acpi_bus_get_device() will always
    return 0 for dock devices in dock_notify(), so the dock station
    docking code under ACPI_NOTIFY_DEVICE_CHECK will never be executed
    and docking will not work as a result of that.

    Fix the problem by making dock_notify() use acpi_device_enumerated()
    to check the presence of the device instead of checking the return
    value of acpi_bus_get_device().

    Fixes: 202317a573b2 (ACPI / scan: Add acpi_device objects for all device nodes in the namespace)
    Signed-off-by: Rafael J. Wysocki <rafael.j.wysocki@intel.com>

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/acpi/dock.c b/drivers/acpi/dock.c
index 8bcfc19..4cbc0bd 100644
--- a/drivers/acpi/dock.c
+++ b/drivers/acpi/dock.c
@@ -611,7 +611,7 @@ static int handle_eject_request(struct dock_station *ds, u32 event)
 static void dock_notify(struct dock_station *ds, u32 event)
 {
  acpi_handle handle = ds->handle;
- struct acpi_device *ad;
+ struct acpi_device *adev = NULL;
  int surprise_removal = 0;
 
  /*
@@ -634,7 +634,8 @@ static void dock_notify(struct dock_station *ds, u32 event)
  switch (event) {
  case ACPI_NOTIFY_BUS_CHECK:
  case ACPI_NOTIFY_DEVICE_CHECK:
-  if (!dock_in_progress(ds) && acpi_bus_get_device(handle, &ad)) {
+  acpi_bus_get_device(handle, &adev);
+  if (!dock_in_progress(ds) && !acpi_device_enumerated(adev)) {
    begin_dock(ds);
    dock(ds);
    if (!dock_present(ds)) {
-- 
1.7.1